### ISSUE TITLE
PyUpgrade rule UP007 do not work on forward reference string types

### ISSUE THREAD

**Issue by goodspark** [astral-sh/ruff#12945]
Keywords: forward reference, pyupgrade, UP007

Python 3.12, also 3.10
Ruff 0.6.1
Command: `ruff check .`

Code:

Python file
```py
x: Optional['Something']  # expect UP007 to trigger here -> x: 'Something | None'
y: Union[int, 'Something']  # expect UP007 to trigger here -> y: 'Something | int'
x2: Optional[Something]  # correctly caught by UP007
y2: Union[int, Something]  # correctly caught by UP007
class Something:
  pass
```

pyproject.toml
```toml
[project]
# As long as this is >=3.10 or any subset, like ==3.10, this bug will occur.
requires-python = ">=3.12"

[tool.ruff.lint]
extend-select = ["UP"]
extend-safe-fixes = ["UP007"]
```

**Comment by charliermarsh** [comment]
I believe that's because `'Something' | None` is not a valid expression.

**Comment by charliermarsh** [comment]
If you add `from __future__ import annotations`, we do correctly flag all four, since the type annotations are then deferred.

**Comment by charliermarsh** [comment]
So I think this is working as intended, AFAICT.

**Comment by goodspark** [comment]
I guess another fix I was envisioning was `"Something | None"` (note both Something and None are in strings).
